### PR TITLE
Integrate `pdet` in GWKokab & Improve PPD Handling and Memory Efficiency

### LINKED ISSUE #288: Tom's Injections Based Neural Net trained VT Implementation in GWKOKAB
### Check for existing issues

- [X] Completed

### Description

We are generating an analytical VT and then training it on NN to apply the detection effects. However better approach is to train the Neural net based on injections which is already implemented and published by Thomas et.al.

So, to get the realistic effects, we need to implement it.
[Tom's Github](https://github.com/tcallister/pdet)

[Paper](https://www.arxiv.org/abs/2408.16828)

[Zenodo](https://zenodo.org/records/13362692)

### LINKED ISSUE #481: feat: add dependencies for gpu/tpu accelerator in `pyproject.toml`
### Check for existing issues

- [x] Completed

### Description

Instead of installing JAX separately for GPU, we should provide an argument like `"gwkokab[gpu]"` to download GPU dependencies in one go.

### PR DESCRIPTION
**Pull request by Qazalbash**
## Summary

This PR contains changes from the following PR,

- #479
- #482
- #483
- #485 
- #486
- #487
- #488
- #489
- #491
- #492
- #493
- #494

## Related To

- closes #288
- #458
- #460
- #461 
- #478
- https://github.com/gwkokab/gwkokab/issues/364#issuecomment-2657613257
- closes #481

### #479

Integrating pdet into GWKokab.

We modified the fork of [pdet](https://github.com/tcallister/pdet) which is available at [gwkokab/pdet](https://github.com/gwkokab/pdet) to work in JAX environment. This PR contains the modification required for pdet to work with existing API styles for population inference, more specifically VT estimation.

We are utilizing [`wcosmo`](https://github.com/ColmTalbot/wcosmo) (#475) as cosmology support for pdet. Author of `wcosmo` has imposed an upper bound on jax version in https://github.com/ColmTalbot/wcosmo/pull/30. Before this PR, we were working with the most recent JAX releases.

### #482

Add platform-specific installation of JAX available in `pyproject.toml`.

We have added support to install platform-specific JAX. Before this, we had to install them separately, occasionally affecting the dependency resolution. Now we can install by specifying `gwkokab[cuda12]` or `gwkokab[tpu]` or `gwkokab` to install for cuda capable GPU, Google TPUs, or CPU respectively. For developer `make install PLATFORM=...`, where `...` will be replaced by the `cuda12` or `tpu` and for CPU based installation `make install` will work.

### #483

We often encounter memory-related bugs. This flag can help us debug some aspect of it.

### #485 

PPD plot generation script for eccentricity matters analysis

### #486

Simplify PPD plot label and range handling.

### #487

With this update, we can pass ranges in the PPD generation script by the name of parameters without any order like `--range name min max step`.

### #488

Weighting final samples by nf model.

### #489

We were facing memory exhaustion errors,

```bash
numpy._core._exceptions._ArrayMemoryError: Unable to allocate 2.33 TiB for an array with shape (200, 200, 200, 200, 200) and data type float64
```

We were making and storing a mesh grid, which is manageable for 1-3 dimensions, but after that, the memory space required to store all this mesh grid goes out of bounds. Therefore, we are not creating meshgrids but iterating over them, which is a little inefficient, but it saves us from the above-mentioned memory errors.

### #491

We have reverted the use of multithreading for ppd generation and returned to JAX-based methods, because they are fast.

Multithreading took enormous time, even days to compute simple and small grid size PPDs. It was better for us to use JAX-based method with little tweaks in memory consumption.

### #492

* Added new arguments for median and PPD line colors, alpha values, and linestyle options.
* Grouped related arguments under a "Pretty Options" section for better organization.
* Updated plotting logic to utilize new customizable parameters for improved visual representation.

### #493

Added a new command-line argument `--font-family` to various plotting scripts, allowing users to specify the font family for rendering text. This enhancement improves customization options for visual outputs across multiple plot types.

### #494

Update `nf_samples_mapping` during log rate wipe to maintain correct indices.

## Summary by CodeRabbit

- **New Features**
  - Enhanced plotting tools now let users customize text rendering via new font and aesthetic options, including extended controls for colors, transparency, and grid styles.
  - The installation process now provides clearer feedback for missing dependencies and integrates an improved release URL.
  - Project packaging has been updated with additional dependencies and new scripts, expanding available modeling parameters.

- **Chores**
  - Version control rules have been refined to include files previously ignored.
  - Optional leak-checking has been added to improve memory management during execution.

### CODE REVIEW

**Comment by Qazalbash** [comment]
@coderabbitai full review

File: src/gwkokab/vts/_pdet/_emulator.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+        # Transform physical parameters to space expected by the neural network
+        transformed_x = self._transform_parameters(
+            *[x[..., i] for i in range(x.shape[-1])]
+        )
+
+        # Apply scaling, evaluate the network, and return
+        scaled_x = (transformed_x - self.scaler["mean"]) / self.scaler["scale"]
+        return jax.vmap(self.nn_vt)(scaled_x)
+
```
[the remark this answers is not in the document]
Comment (zeeshan5885) [reply src/gwkokab/vts/_pdet/_emulator.py:167]: @Qazalbash  Any thought son this?

File: src/kokab/utils/ppd.py [hunk 2]
```diff
@@ -162,45 +213,56 @@ def save_probs(
...
+    parameters : List[str]
+        The parameters of the model.
+    constants : Dict[str, ArrayLike]
+        The constants of the model.
+    nf_samples_mapping : Dict[str, int]
+        The mapping of the normalizing flow samples to the model parameters.
+    batch_size : int, optional
+        The batch size for the computation, defaults to 1000.
+    """
+    xx = [jnp.linspace(a, b, int(n)) for a, b, n in domains]
```
[the remark this answers is not in the document]
Comment (zeeshan5885) [reply src/kokab/utils/ppd.py:245]: @Qazalbash  Please have a look into it.

File: src/gwkokab/vts/_pdet/_emulator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Qazalbash) [reply src/gwkokab/vts/_pdet/_emulator.py:167]: @zeeshan5885 we can experiment with nested `jax.lax.map`s, because this function is being called from a `jax.lax.map`.
